server/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java

```diff
@@ -513,6 +528,122 @@ private IndexItemResult indexItem(Indexer indexer,
         throw lastException;
     }
 
+    @Nullable
+    private IndexItemResult indexItemForUpdate(Indexer indexer,
```

Comment:
I get the impression that there's now quite a bit of duplication here between update/insert.

Isn't it possible to share the logic once there's a `ParsedDocument` ? The indexing logic after that must be the same.